### PR TITLE
http2: protocol-level client specs

### PR DESCRIPTION
**Pull request by raboof**
Infrastructure for 'protocol-level' tests for the client side: allows testing the translation from request object to network frames, and from response network frames to response object.

Heavily inspired by `Http2ServerSpec`, and sharing some of the code.

Kept most of `Http2ServerSpec` as commented-out `Http2ClientSpec`, as for some of those it might make sense to include the 'mirrored' scenario.

### CODE REVIEW

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/framing/Http2FrameParsing.scala [hunk 1]
```diff
@@ -39,6 +40,113 @@ private[http] object Http2FrameParsing {
     readSettings(Nil)
   }
 
+  def parseFrame(tpe: FrameType, flags: ByteFlag, streamId: Int, payload: ByteReader, log: LoggingAdapter): FrameEvent = {
```
Comment (raboof) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/framing/Http2FrameParsing.scala:43]: (moved here without changes apart from passing in the 'log' parameter)

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2FrameHpackSupport.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import com.twitter.hpack.{ Decoder, Encoder, HeaderListener }
+import org.scalatest.concurrent.ScalaFutures
+
+import scala.collection.immutable.VectorBuilder
+
+/** Helper that allows automatic HPACK encoding/decoding for wire sends / expectations */
+trait Http2FrameHpackSupport extends Http2FrameProbeDelegator with Http2FrameSending with ScalaFutures {
```
Comment (raboof) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2FrameHpackSupport.scala:17]: Moved here without significant changes from `AutomaticHpackWireSupport` in `Http2ServerSpec`

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2FrameSending.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package akka.http.impl.engine.http2
+
+import java.nio.ByteOrder
+
+import akka.http.impl.engine.http2.FrameEvent.{ ContinuationFrame, HeadersFrame, PriorityFrame, Setting, SettingsFrame, WindowUpdateFrame }
+import akka.http.impl.engine.http2.Http2Protocol.{ ErrorCode, Flags, FrameType, SettingIdentifier }
+import akka.http.impl.engine.http2.framing.FrameRenderer
+import akka.util.{ ByteString, ByteStringBuilder }
+
+trait Http2FrameSending {
```
Comment (raboof) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2FrameSending.scala:10]: Extracted from `TestSetupWithoutHandshake` in `Http2ServerSpec`.


While the `TestSetupWithoutHandshake` itself is different between client and server, the utilities for creating and sending frames can be shared (though they will represent incoming 'request frames' for the server test and incoming 'response frames' for the client test)

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2FrameProbe.scala [hunk 4]
```diff
@@ -106,6 +110,18 @@ private[http] object Http2FrameProbe extends Matchers {
       def expectNoBytes(): Unit = probe.expectNoBytes()
       def expectNoBytes(timeout: FiniteDuration): Unit = probe.expectNoBytes(timeout)
 
+      def expectFrame(): FrameEvent = {
+        // Not supporting large frames or high streamId's here for now, throw when we encounter those.
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2FrameProbe.scala:114]: :+1:

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+              .withEntity(Strict(ContentTypes.NoContentType, ByteString.empty))
+        )
+
+        emitRequest(3, HttpRequest(uri = "http://www.example.com/"))
+        expectFrame() shouldBe a[HeadersFrame]
+
+        val incorrectHeaderBlock = hex"00 00 01 01 05 00 00 00 01 40"
+        sendHEADERS(4, endStream = true, endHeaders = true, headerBlockFragment = incorrectHeaderBlock)
+
+        // TODO shouldn't this be '3'?
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:120]: Ultimately, it should be but that flag is best effort and our best effort is currently quite low...

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 6]
```diff
@@ -1060,39 +1064,14 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
...
-      implicit val bigEndian: ByteOrder = ByteOrder.BIG_ENDIAN
-      val bb = new ByteStringBuilder
-      bb.putInt(errorCode.id)
-      sendFrame(FrameType.RST_STREAM, ByteFlag.Zero, streamId, bb.result())
+      super.sendDATA(streamId, endStream, data)
     }
 
-    def sendWINDOW_UPDATE(streamId: Int, windowSizeIncrement: Int): Unit = {
-      sendBytes(FrameRenderer.render(WindowUpdateFrame(streamId, windowSizeIncrement)))
+    override def sendWINDOW_UPDATE(streamId: Int, windowSizeIncrement: Int): Unit = {
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:1073]: The window stuff is the same on client and server side so it might make sense to extract that as well. But this can maybe be done in another PR.

**Review by jrudolph — APPROVED** [review]
LGTM
